### PR TITLE
Add Kavita extension

### LINKED ISSUE #9332: Kavita
### Source name

Kavita

### Source link

https://kavitareader.com/

### Language

English

### Other details

The api is still being built upon. (Swagger ui) 
https://wiki.kavitareader.com/en/guides/plugins-and-3rd-party-scripts
Needs to be hosted first. if u need more requirements u can ask dev Joe at their [Discord ](https://discord.gg/hy73F2Dk). Thanks
https://api.kavitareader.com/
 Alternative approach, scrap kavita if api isn't enough. Yea I know mad. But the authentication panel won't change nor the sending protocol so #8768 

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a title with source name.
- [X] I have checked that the extension does not already exist on the [website extensions list](https://tachiyomi.org/extensions/) or the app.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/tachiyomiorg/tachiyomi-extensions/) and verified it does not appear in the code base.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by ThePromidius**
### Description:
Supports all services provided by kavita API.
Expected to add two-way tracking in the future
Minimum Kavita version: v0.4.9.24
 
Closes #9332 

### Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

Disclaimer:
I still don't control github very well so apologies if anything is wrong. Tips and comments are welcomed.

More info on the extension:
https://wiki.kavitareader.com/en

### CODE REVIEW

File: src/all/kavita/build.gradle [hunk 1]
```diff
@@ -0,0 +1,16 @@
+apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlinx-serialization'
+
+ext {
+    extName = 'Kavita'
+    pkgNameSuffix = 'all.kavita'
+    extClass = '.Kavita'
+    extVersionCode = 6
```
Comment (alessandrojean) [inline src/all/kavita/build.gradle:9]: Extension versions should start in 1 if they are new.

```suggestion
    extVersionCode = 1
```

File: src/all/kavita/src/eu/kanade/tachiyomi/extension/all/kavita/Kavita.kt [hunk 2]
```diff
@@ -0,0 +1,1036 @@
...
+        val apiKey = tokens[1]
+        val baseUrlSetup = tokens[0].replace("\n", "\\n")
+        if (!baseUrlSetup.startsWith("http")) {
+            throw Exception("""Url does not start with "http/s" but with ${baseUrlSetup.split("://")[0]} """)
+        }
+        preferences.edit().putString("APIKEY", apiKey).commit()
+        preferences.edit().putString("BASEURL", baseUrlSetup).commit()
+
+        val body = JSONObject().toString()
+            .toRequestBody("application/json; charset=utf-8".toMediaTypeOrNull())
```
Comment (alessandrojean) [inline src/all/kavita/src/eu/kanade/tachiyomi/extension/all/kavita/Kavita.kt:772]: Can probably just hardcode the string and remove the usage of `org.json` stuff.

```suggestion
        val body = "{}".toRequestBody("application/json; charset=utf-8".toMediaTypeOrNull())
```

**Review by alessandrojean — CHANGES_REQUESTED** [review]

**Review by alessandrojean — APPROVED** [review]

**Comment by jopejoe1** [comment]
Please uses these icons to be more consistent with other extensions.
[ic_launcher(6).zip](https://github.com/tachiyomiorg/tachiyomi-extensions/files/7777699/ic_launcher.6.zip)

**Comment by ThePromidius** [comment]
> Please uses these icons to be more consistent with other extensions. [ic_launcher(6).zip](https://github.com/tachiyomiorg/tachiyomi-extensions/files/7777699/ic_launcher.6.zip)

Much appreciated. TBH i had problems finding some good icons.

**Comment by ThePromidius** [comment]
Marked as draft till there is a stable build version of kavita that supports the extension.

**Comment by ThePromidius** [comment]
> Please uses these icons to be more consistent with other extensions. [ic_launcher(6).zip](https://github.com/tachiyomiorg/tachiyomi-extensions/files/7777699/ic_launcher.6.zip)

Would it be possible to have those icons but changing the black border with the same green the circle is made of?
Would that still be consistent with the rest?

**Comment by jopejoe1** [comment]
> > Please uses these icons to be more consistent with other extensions. [ic_launcher(6).zip](https://github.com/tachiyomiorg/tachiyomi-extensions/files/7777699/ic_launcher.6.zip)
> 
> Would it be possible to have those icons but changing the black border with the same green the circle is made of? Would that still be consistent with the other extensions?

Yes that would still be consistent
[ic_launcher(7).zip](https://github.com/tachiyomiorg/tachiyomi-extensions/files/7886071/ic_launcher.7.zip)

**Comment by ThePromidius** [comment]
Kavita v0.5.0 (stable) is ready. PR ready for revision.

**Review by alessandrojean — APPROVED** [review]
The URL checking is just a suggestion, don't need to be followed. The other things looks fine to me :+1:.

**Comment by majora2007** [comment]
Any chance we can get this merged in? I see we have approval from a member of Tachiyomi.

**Comment by edgehub-cloud** [comment]
Bllesed

File: src/all/kavita/src/eu/kanade/tachiyomi/extension/all/kavita/Kavita.kt [hunk 3]
```diff
@@ -952,7 +977,11 @@ class Kavita(suffix: String = "") : ConfigurableSource, HttpSource() {
         val baseUrlSetup = tokens[0].replace("\n", "\\n")
 
         if (!baseUrlSetup.startsWith("http")) {
```
[the remark this answers is not in the document]
Comment (ThePromidius) [reply src/all/kavita/src/eu/kanade/tachiyomi/extension/all/kavita/Kavita.kt:994]: thanks will do

**Pull request merged**